realm/realm-library/src/main/cpp/objectserver_shared.hpp

```diff
@@ -61,27 +63,58 @@ class JniSession {
                     error_code.category() != realm::sync::client_error_category()) {
                 // FIXME: Consider below when moving to the OS sync manager.
                 // Ignore this error since it may cause exceptions in java ErrorCode.fromInt(). Throwing exception there
                 // will trigger "called with pending exception" later since the thread is created by java, and the
                 // endless loop is in native code. The java exception will never be thrown because of the endless loop
                 // will never quit to java land.
-                realm::jni_util::Log::e("Unhandled sync client error code %1, %2. is_fatal: %3.",
-                                        error_code.value(), error_code.message(), is_fatal);
+                realm::jni_util::Log::e("Unhandled sync client error code %1, %2. is_fatal:", error_code.value(), error_code.message(), is_fatal);
                 return;
             }
 
-            auto session_ref = weak_session_ref.lock();
-            if (session_ref) {
-                session_ref.get()->call_with_local_ref([&](JNIEnv* local_env, jobject obj) {
-                    static realm::jni_util::JavaMethod notify_error_handler(
-                            local_env, obj, "notifySessionError", "(ILjava/lang/String;)V");
-                    local_env->CallVoidMethod(
-                            obj, notify_error_handler, error_code.value(), local_env->NewStringUTF(message.c_str()));
+            // Handle client reset, without returning to Java
+
+            // we don't have the original SyncError so we can't call SyncError#is_client_reset_requested
+            // we need to transform the error code to an enum, then do the check manually
+            using ProtocolError = realm::sync::ProtocolError;
+            auto protocol_error = static_cast<ProtocolError>(error_code.value());
+
+            // Documented here: https://realm.io/docs/realm-object-server/#client-recovery-from-a-backup
+            if (protocol_error == ProtocolError::bad_server_file_ident
+                || protocol_error == ProtocolError::bad_client_file_ident
+                || protocol_error == ProtocolError::bad_server_version
+                || protocol_error == ProtocolError::diverging_histories) {
+
+                // Add a SyncFileActionMetadata marking the Realm as needing to be deleted.
+                auto recovery_path = realm::util::reserve_unique_file_name(
+                        realm::SyncManager::shared().recovery_directory_path(),
+                        realm::util::create_timestamped_template("recovered_realm"));
+                auto original_path = local_realm_path;
+
+                realm::SyncManager::shared().perform_metadata_update([original_path = std::move(original_path),
+                        recovery_path = std::move(recovery_path)](const auto &manager) {
+                    realm::SyncFileActionMetadata(manager,
```

Comment:
It writes the action to the metadata table, and when the app restarts, the action will be performed. To do the reset here, `immediate_run_file_actions()` should be called.